Detectors/PHOS/reconstruction/src/CaloRawFitterGS.cxx

```diff
@@ -108,208 +119,172 @@ CaloRawFitterGS::FitStatus CaloRawFitterGS::evalFit(gsl::span<short unsigned int
-            At += s;
-            Bt += s * xiprev;
-            Ct += s * xiprev * xiprev;
-            y2 += (atmp * atmp - ap * ap);
-          }
-        } else {
-          app = ap;
-          ap = a;
-        }
-
-        //to calculate time
-        float ss = a * ma0[i];
-        At += ss;
-        Bt += ss * xi;
-        Ct += ss * xi * xi;
-        y2 += a * a;
-      }
-      //evaluate new time and amp
+  // calculate time, amp and chi2
+  double a, b, c, d, e; // Polinomial coefficients
+  if (!mOverflow) {
+    a = ma1[j] * b0 - ma0[j] * b1;
+    b = ma0[j] * b2 + 2. * ma1[j] * b1 - 3. * ma2[j] * b0;
+    c = 3. * (ma3[j] * b0 - ma1[j] * b2);
+    d = 3. * ma2[j] * b2 - ma4[j] * b0 - 2. * ma3[j] * b1;
+    e = ma4[j] * b1 - ma3[j] * b2;
+  } else { // account removed points in overflow
+    a = (ma1[j] - sa1) * b0 - (ma0[j] - sa0) * b1;
+    b = (ma0[j] - sa0) * b2 + 2. * (ma1[j] - sa1) * b1 - 3. * (ma2[j] - sa2) * b0;
+    c = 3. * ((ma3[j] - sa3) * b0 - (ma1[j] - sa1) * b2);
+    d = 3. * (ma2[j] - sa2) * b2 - (ma4[j] - sa4) * b0 - 2. * (ma3[j] - sa4) * b1;
+    e = (ma4[j] - sa4) * b1 - (ma3[j] - sa3) * b2;
+  }
 
-      double polB = At - Bt;
-      double polC = Ct - 2. * Bt;
-      if (At == 0.) {
-        if (polB == 0.) {
-          mTime = 999.;
-          status = kFitFailed;
-        } else {
-          mTime = -polC / (2. * polB);
-        }
-      } else {
-        double d = polB * polB - At * polC;
-        if (d >= 0) {
-          mTime = (-polB - sqrt(d)) / (At * mDecTime);
-        } else {
-          mTime = 999.;
-          status = kFitFailed;
-        }
-      }
-      if (status == kFitFailed) { //Can not improve, give up
-        mAmp = 0;
-        mTime = 999.;
-        mChi2 = 999.;
-        mOverflow = false;
-        return status;
-      }
-      double tt = mTime * mDecTime;
-      expT = exp(tt);
-      double nom = (At * tt * tt - 2. * Bt * tt + Ct) * expT; //  1./(k*k) cancel with denom
-      int i = nSamples - 1;
-      double denom = ((mb4[i] - tt * (mb3[i] - tt * (mb2[i] - tt * (mb1[i] - tt * mb0[i]))))) *
-                     expT * expT / (mDecTime * mDecTime); // 1/(k*k) cancel with nom
-      if (denom != 0) {
-        mAmp = nom / denom;
-        mChi2 = (y2 - (2. * (At * tt * tt - 2. * tt * Bt + Ct) - mAmp * denom * expT) * mAmp * expT / (mDecTime * mDecTime)) / (i + 1);
+  // Find zero of 4-order polinomial
+  // first use linear extrapolation to reach correct root of four
+  double z = -1.;
+  if (ma0[j] * b1 - ma1[j] * b0 != 0) {
+    z = (ma1[j] * b1 - ma2[j] * b0) / (ma0[j] * b1 - ma1[j] * b0) - 1.; // linear fit + offset
+  }
+  double q = 0., dq = 0., ddq = 0., lq = 0., dz = 0.1;
+  q = a * z * z * z * z + b * z * z * z + c * z * z + d * z + e; // polinomial
+  dq = 4. * a * z * z * z + 3. * b * z * z + 2. * c * z + d;     // Derivative
+  ddq = 12. * a * z * z + 6. * b * z + 2. * c;                   // Second derivative
```

Comment:
you could decrease the number of multiplications by precalculated quantities.

Reply:
Hi @shahor02, thank you, I modified code. As for math - I checked, method reproduces fitting with Minuit with good accuracy.